denizen_scripts/survival/repo-link/items/system/soul_item_system.dsc

```diff
@@ -247,58 +337,78 @@ item_with_soul:
   material: diamond_sword
   display name: <&c>ERROR - REPORT THIS
 
+#@get_random_soul:
+#@  type: procedure
+#@  definitions: rarity|level
+#@  script:
+#^    - define soul_type <script[item_system_global_data].list_keys[soul_names.<[rarity]>].random>
+#^    - foreach <script[item_system_global_data].yaml_key[soul_names.<[rarity]>.<[soul_type]>]> as:modifier:
+#^      - define stat <[modifier].before[/]>
+#^      - define type <[modifier].after[/]>
+#^      - if <[type]> == buff:
+#^        - define buffs:->:<[stat]>,<script[item_system_global_data].yaml_key[calculations.<[stat]>].parsed>
+#^      - else:
+#^        - define debuffs:->:<[stat]>,<script[item_system_global_data].yaml_key[calculations.<[stat]>].parsed>
+#^    - if <[buffs]||null> != null:
+#^      - define buffs_and_debuffs:|:buffs/<[buffs].escaped>
+#^    - if <[debuffs]||null> != null:
+#^      - define buffs_and_debuffs:|:debuffs/<[debuffs].escaped>
+#^    - define flavor:<element[<&d>Soul&spItem<&nl><&e>Combine&spwith&spArmor&spor&spWeapons].escaped>
+#^    - determine <proc[item_system_build_item].context[<item[soul].with[nbt=soul/<[soul_type]>,<[level]>|rarity/<[rarity]>|soul_level/<[level]>|active_soul/<[soul_type]>|flavor/<[flavor]>|<[buffs_and_debuffs].separated_by[|]>]>]>
+
 get_random_soul:
   type: procedure
   definitions: rarity|level
   script:
     - define soul_type <script[item_system_global_data].list_keys[soul_names.<[rarity]>].random>
-    - foreach <script[item_system_global_data].yaml_key[soul_names.<[rarity]>.<[soul_type]>]> as:modifier:
-      - define stat <[modifier].before[/]>
-      - define type <[modifier].after[/]>
+    - define NBT <map.with[soul].as[<map.with[type].as[<[soul_type]>].with[level].as[<[level]>]>]>
+    - define NBT <[NBT].with[rarity].as[<[rarity]>].with[soul_level].as[<[level]>].with[active_soul].as[<[soul_type]>]>
+    - foreach <script[item_system_global_data].data_key[soul_names.<[rarity]>.<[soul_type]>]> Key:Stat as:Type:
```

Comment:
initially, this looped across a hacked `key/value` list map, in the example for a soul item named "bear" with a rarity of "1":
`armor/buff|attack_speed/buff|health/debuff`
now loops across the mapping:
```yml
bear:
  health: buff
  melee_damage: buff
  attack_speed: debuff
```